generator/internal/http_annotation_parser_test.cc

```diff
@@ -0,0 +1,262 @@
+                                             {
+                                                 MakeSegment("projects"),
+                                                 MakeMatchSegment(),
+                                                 MakeSegment("databases"),
+                                                 MakeMatchSegment(),
+                                                 MakeSegment("documents"),
+                                                 MakeMatchSegment(),
+                                                 MakeMatchRecursiveSegment(),
+                                             }}),
+          MakeSegment(PathTemplate::Variable{/*field_path=*/"collection_id",
+                                             /*segments=*/{}}),
+      },
+      /*verb=*/{}};
+  EXPECT_TRUE(SameValues(expected, *parsed))
+      << ", expected=" << expected << ", parsed=" << *parsed;
+}
+
+TEST(ParseHttpTemplate, WithVerb) {
+  auto parsed = ParsePathTemplate("/v1/{project}:put");
+  ASSERT_STATUS_OK(parsed);
+  auto const expected = PathTemplate{
+      /*segments=*/{
+          MakeSegment("v1"),
+          MakeSegment(PathTemplate::Variable{/*field_path=*/"project",
+                                             /*segments=*/{}}),
+      },
+      /*verb=*/"put"};
+  EXPECT_TRUE(SameValues(expected, *parsed))
+      << ", expected=" << expected << ", parsed=" << *parsed;
+}
+
+TEST(ParseHttpTemplate, Errors) {
+  struct {
+    std::string input;
+    std::string expected;
+  } const cases[] = {
+      {"v1/projects/{project}", " offset 0\n"},
+      {"/", " offset 1\n"},
+      {"/:put", " offset 1\n"},
+      {"/v1:bad|verb", " offset 7\n"},
+      {"/v1:put ", " offset 7\n"},
+      {"/v1//", " offset 4\n"},
+      {"/v1/|**/", " offset 4\n"},
+      {"/v1/a/{p|}", " offset 8\n"},
+      {"/v1/a/{p=", " offset 9\n"},
+      {"/v1/a/{p=}", " offset 9\n"},
+      {"/v1/a/{=abc}", " offset 7\n"},
+      {"/v1/a/{p=**|}", " offset 11\n"},
+      {"/v1/a/{p", " offset 8\n"},
+      {"/v1/a/{p} ", " offset 9\n"},
+      {"/v1/a/{p=b/{c}}", " offset 11\n"},
+      {"/v1/a/{", " offset 7\n"},
+  };
+  for (auto const& test : cases) {
+    SCOPED_TRACE("Testing with input=" + test.input);
+    auto parsed = ParsePathTemplate(test.input);
+    EXPECT_THAT(parsed, StatusIs(StatusCode::kInvalidArgument,
+                                 HasSubstr(test.expected)));
+  }
+}
+
+TEST(ParsetHttpTemplate, OStream) {
+  auto parsed = ParsePathTemplate("/v1/a/{b=c}/d/{e=f/**/*}:put");
+  ASSERT_STATUS_OK(parsed);
+  std::ostringstream os;
+  os << *parsed;
+  EXPECT_EQ(os.str(),
+            "{segments=[ {v1} / {a} / {field_path=b, segments=[ {c} ]} / {d} /"
+            " {field_path=e, segments=[ {f} / {**} / {*} ]} ], verb=put}");
+}
+
+}  // namespace
+}  // namespace generator_internal
+}  // namespace cloud
+}  // namespace google
+
+int main(int argc, char* argv[]) {
+  ::testing::InitGoogleTest(&argc, argv);
+  return RUN_ALL_TESTS();
+}
```

Comment:
Observation:  It's a shame the generator tests don't use `gtest_main` as it doesn't look like any of them require special treatment.

Reply:
Agreed.